### DISCUSSION TITLE
Working with LOB objects

### DISCUSSION THREAD

**Discussion by vikingUnet** [psycopg/psycopg#38] [General]
Hello!
In psycopg2 to insert LOB we should used function "lo_from_bytea(0, %s)" (wich gets only oid of LOB) and to read LOB we used "lo_get(oid)". Can we simpy send opened file stream to insert and select? It will be more frendly feauter in psycopg3 version. Thanks!

**Comment by dvarrazzo** [comment]
Hello,

currently psycopg2 uses the libpq functions to manage the objects: https://www.postgresql.org/docs/current/lo-interfaces.html

However these don't support an async interface: they are only server-side function called using the blocking `PQfn()` function. So for psycopg3 I would rather call the server-side functions directly.

I understand you would like to pass a file stream to copy a file into a LO. This seems rather the way copy works in psycopg2. My experience of using this interface is that it's a bad one to create data dynamically: if you have a data generator which can push data to a file you have to create a sort of background wrapper to have copy pulling data from it: when I had to do it it was very awkward :( Furthermore a Python file object has a blocking interface: you couldn't use this way of working in a blocking program.

In psycopg3 copy doesn't work with a file interface anymore: it has `read()` and `write()` functions which are async for an async connection, so the loop on the block is left to the user: https://www.psycopg.org/psycopg3/docs/copy.html#copying-block-by-block I would use a similar interface for LO too.

@vikingUnet Could you please provide an example of how you think it would be a good way of using large objects? Thank you
